### ISSUE TITLE
BFS7.7.1 Failing Jenkins Job Issue

### ISSUE THREAD

**Issue by 2samueld** [Tengda-He/kibana#57]
# Description:

Currently Jenkins jobs for bfs7.71 randomly fails. Below is identified issue: 
 
1. issue with optimizing bundles. 
      - Jenkins job 15 and 20 https://jenkins.bfs.sichend.people.aws.dev/blue/organizations/jenkins/Kibana/detail/bfs7.7.1/15/pipeline 
2. Jenkins switch node version from 10.19.0 to 8.14.0
      - `Kibana does not support the current Node.js version v8.14.0. Please use Node.js v10.19.0.`
3. Functional tests fail due to element not located by webdriver ( jenkins Job 16)
      - Chrome UI Functional Tests.test/functional/apps/console/_console·ts.console app console app "before all" hook for "should show the default request"
      - Chrome UI Functional Tests.test/functional/apps/context/index·js.context app "before all" hook in "context app"
      - Chrome UI Functional Tests.test/functional/apps/status_page/index·js.status page "before each" hook for "should show the kibana plugin as ready"
      - Chrome UI Functional Tests.test/functional/apps/timelion/_expression_typeahead·js.timelion app expression typeahead "before all" hook for "should display function suggestions filtered by function name"

**Comment by Tengda-He** [comment]
Lets focus on addressing these problems, this is top priority, thanks

**Comment by 2samueld** [comment]
### Update on Above Issue 2 & 3:

2. **Node environment** has been fixed with PR #59 
3. **Element not located by webdriver** never appeared again. Multiple jenkins job has been run since then and the functional tests always passes

**Comment by 2samueld** [comment]
### Analysis on Issue 1:

**Problem:** 
Optimization fails due to modules not being found. These modules (such as timelion and statusPage) are located in the `kibana/optimize/bundles` folder. When the script `node scripts/kibana` is ran in CiGroups, the following process happens:
1. Loop through every compilation module to start building
2. Compare the new build entry path with the previous entry path
3. Run dll compilation (optimization) in case the previous build doesn't have an entry path or if the new build entry path differs from the previous entry path

**Diagnosis:**
Logs were added in three functions `registerBeforeCompileHook, registerCompilationHook, and registerDoneHook` found in  `kibana/src/optimize/dynamic_dll_plugin/dynamic_dll_plugin.js` file.
- **registerBeforeCompileHook:** this function is called when dll plugin tasks starts. It ensures modules are registered before compilation. A log was added to track whether all the modules were registered.
- **registerCompilationHook:** this is the function that is called when `node scripts/kibana` is ran as described above. Logs were added to check if the previous and new build entry paths are the same, if compilation is required, and if dll bundles exists.
- **registerDoneHook:** this function ensures compilation checks are finished. Logs were added to check whether if there were any modules that needed compilation but were omitted.

**Comment by seraphjiang** [comment]
what does random means?

beside ciGroup5, did we see any other ciGroup random failure?
what's different between failed ciGroup5 and other passed ciGroup?

did we even pass the ciGroup5? is it passed local run?

**Comment by 2samueld** [comment]
@seraphjiang in the below job, CiGroup 9 fails because of optimization failure. The difference between the failed CiGroup and passed CiGroup is an optimization issue, where some modules in `optimize/bundles` directory are not found.

https://jenkins.bfs.sichend.people.aws.dev/blue/organizations/jenkins/Kibana/detail/bfs7.7.1/20/pipeline/86

**Comment by seraphjiang** [comment]
how about run single ciGroup separately, could we always pass the tests?
if so, could we try run ciGroup in sequentially instead of in parallel?

**Comment by 2samueld** [comment]
Running tests separately/sequentially takes a couple of hours to finish functional tests compared to running it parallel, which takes about half an hour to 45 mins. 

Coming to the question of " will running CiGroups separately always pass the tests?". There isn't enough data to show that it'll pass. A couple of Jenkins job that ran CiGroup separately will be started this evening. That can give us a picture of the test results.

**Comment by sichend** [comment]
I have spent some time from Friday to research on this issue

FATAL  Error: Optimizations failure.
  ERROR in ./optimize/bundles/timelion.entry.js
  Module build failed (from ./node_modules/thread-loader/dist/cjs.js):
    Thread Loader (Worker 0)
    ENOENT: no such file or directory, open '/var/lib/jenkins/workspace/Kibana_bfs7.7.1/optimize/bundles/timelion.entry.js'

The cause of this is the combination of the two factors, the optimization process during Kibana startup and the Jenkins pipeline docker plugin. Please see more details below

Background knowledge:
1. Kibana startup. With the current configuration in the functional testing, Kibana goes through the optimize stage during each start up. The [optimizer](https://github.com/Tengda-He/kibana/blob/bfs7.7.1/src/optimize/fs_optimizer.js) overwrites the files in the root workspace `*/optimize/bundles/*` with the new compiled plugins.
2. When Jenkins pipeline runs commands in a docker environment, it mount the current project workspace into the container as the workspace with read and write permission. This mount type persist any changes in the container, i.e. caches from the build process, into the host file system

Problem Root Cause:
Let's see the docker commands to parallel execute the functional test

`
// CI group 1
docker run -t -d -u 111:115 -w /var/lib/jenkins/workspace/Kibana_bfs7.7.1 -v /var/lib/jenkins/workspace/Kibana_bfs7.7.1:/var/lib/jenkins/workspace/Kibana_bfs7.7.1:rw,z -v /var/lib/jenkins/workspace/Kibana_bfs7.7.1@tmp:/var/lib/jenkins/workspace/Kibana_bfs7.7.1@tmp:rw,z  bfs7.7.1-test-image:108 cat

docker top 03ba0a2e99baeea696fbe9bbe1168cc2b199a273cb9a8ccf4a6dd1c6c0c1f22c -eo pid,comm

docker exec -it 03ba0a2e99baeea696fbe9bbe1168cc2b199a273cb9a8ccf4a6dd1c6c0c1f22c  "node scripts/functional_tests.js --config test/functional/config.js --include ciGroup1:

// CI group 2
docker run -t -d -u 111:115 -w /var/lib/jenkins/workspace/Kibana_bfs7.7.1 -v /var/lib/jenkins/workspace/Kibana_bfs7.7.1:/var/lib/jenkins/workspace/Kibana_bfs7.7.1:rw,z -v /var/lib/jenkins/workspace/Kibana_bfs7.7.1@tmp:/var/lib/jenkins/workspace/Kibana_bfs7.7.1@tmp:rw,z bfs7.7.1-test-image:108 cat
docker top 76a08b05326c3f895a8ec75ef8fa032d8ae303648e5602657ea0c9fa59b7dbc9 -eo pid,comm

docker exec -it 03ba0a2e99baeea696fbe9bbe1168cc2b199a273cb9a8ccf4a6dd1c6c0c1f22c  "node scripts/functional_tests.js --config test/functional/config.js --include ciGroup2:
`

Now, we can clearly see that the Jenkins Pipeline is trying to mount the same workspace to multiple parallel containers with read/write permissions at the same time. Each parallel container starts up a Kibana instance and goes through the optimization process, overwriting the same file. This creates a race condition among all the parallel containers during the Kibana optimization process. Hence, we see the plugin file not found errors, with unpredictable files each time. 

How to solve this problem?
As you can see, there are two reasons to cause this issue. Therefore, fixing either of them should resolve the flaky test issue. Here are some suggestions, yet to be tested. 
1. From Jenkins side, you can probably fix this by duplicate the workspace for each parallel containers, specify the volume mount for each containers. Here is the documentation to start with, [link](https://www.jenkins.io/doc/book/pipeline/docker/)
2. From Kibana side, disable the optimization should solve the problem. However, I will let you experts to decide if this is an acceptable approach for this issue. 

Please let me know if you have any questions. Thanks.

**Comment by sichend** [comment]
In order to speed up the progress to remove this blocker, I have done some testing myself last night, I would like to put some test results. I have tested three different approaches. Please see details below
1. Disabling optimization by changing the test Kibana environment configuration. This ended up to be unsuccessful. Considering the intricacy of the Kibana system and startup process, this might not be a good way to move forward.
2. Changing the working directory of the Kibana environment inside docker containers. There are two approaches.
    - Make a copy of the current Jenkins workspace for each parallel container, mount to each container, run kibana environment, clean up the workspace copies. This approach is quick and dirty. Since the entire Kibana code base is fairly large, copying the directory for 12 times consumes fairly amount of time and disk space. There is the test run [link](https://jenkins.bfs.sichend.people.aws.dev/job/Kibana/job/bfs7.7.1/140/), total run time if 37 minutes. Here is the sample code for this approach 
    ```
            stage("${currentCiGroup}") {
                sh "rm -rf ${env.WORKSPACE}_${currentCiGroup}"
                sh "cp -r ${env.WORKSPACE} ${env.WORKSPACE}_${currentCiGroup}"
                
                withEnv([
                    "TEST_BROWSER_HEADLESS=1",
                    "CI=1",
                    "CI_GROUP=${currentCiGroup}",
                    "GCS_UPLOAD_PREFIX=fake",
                    "TEST_KIBANA_HOST=localhost",
                    "TEST_KIBANA_PORT=6610",
                    "TEST_ES_TRANSPORT_PORT=9403",
                    "TEST_ES_PORT=9400",
                    "CI_PARALLEL_PROCESS_NUMBER=${currentStep}",
                    "JOB=ci${currentStep}",
                    "CACHE_DIR=${currentCiGroup}"
                ]) {
                    testImage.inside("-v \"${env.WORKSPACE}_${currentCiGroup}:${env.WORKSPACE}_${currentCiGroup}\""){
                        sh "cd ${env.WORKSPACE}_${currentCiGroup}"
                        sh "node scripts/functional_tests.js --config test/functional/config.js --include ${currentCiGroup}"
                    }
                }
                sh "rm -rf ${env.WORKSPACE}_${currentCiGroup}"
            }
    ```
   
    - The second approach is an improvement based on the first. Considering all the racing condition happens only in the */optimize/ directory, we can simply ignore all other directory. We only need to create an empty directory and mount into each container as the */optimize/ directory in the workspace. In this case, all the containers are sharing everything but the /optimize/ directory. This approach avoids the inefficiency caused by duplicating workspaces. This run completes in 24 minutes comparing to the previous approach. There is the run [link](https://jenkins.bfs.sichend.people.aws.dev/job/Kibana/job/bfs7.7.1/141/) and test code. 
    ```
    stage("${currentCiGroup}") {
                sh "rm -rf ${env.WORKSPACE}_${currentCiGroup}"
                sh "mkdir ${env.WORKSPACE}_${currentCiGroup}"
                
                withEnv([
                    "TEST_BROWSER_HEADLESS=1",
                    "CI=1",
                    "CI_GROUP=${currentCiGroup}",
                    "GCS_UPLOAD_PREFIX=fake",
                    "TEST_KIBANA_HOST=localhost",
                    "TEST_KIBANA_PORT=6610",
                    "TEST_ES_TRANSPORT_PORT=9403",
                    "TEST_ES_PORT=9400",
                    "CI_PARALLEL_PROCESS_NUMBER=${currentStep}",
                    "JOB=ci${currentStep}",
                    "CACHE_DIR=${currentCiGroup}"
                ]) {
                    testImage.inside("-v \"${env.WORKSPACE}_${currentCiGroup}:${env.WORKSPACE}/optimize\""){
                        sh "node scripts/functional_tests.js --config test/functional/config.js --include ${currentCiGroup}"
                    }
                }
                sh "rm -rf ${env.WORKSPACE}_${currentCiGroup}"
            }
    ```

My conclusion is that the second approach to change the Jenkins file is a cleaner, more efficient, and least changes to the original test cases and environment. We can consider to use this approach. However, please take a challenge to improve this approach by making it even cleaner. In my test code, I created a directory outside the current Jenkins workspace. This is a quick and dirty way to PoC. We should contain everything within the current Jenkins workspace. Therefore, a better approach is to create a directory ${env.WORKSPACE}/.optimize/${currentCiGroup}, and mount this directory into each container. Eventually remove these temp directories. 

Please let me know if you have questions.

**Comment by 2samueld** [comment]
> Fix for parallel racing functional test #80

Applying the following second approach, functional parallel tests passes. 
```
ciGroupsMap["${currentCiGroup}"] = {
            sh "rm -rf ${env.WORKSPACE}/.optimize/${currentCiGroup}"
            sh "mkdir -p ${env.WORKSPACE}/.optimize/${currentCiGroup}"
            stage("${currentCiGroup}") {
                withEnv([
                    "TEST_BROWSER_HEADLESS=1",
                    "JOB=ci${currentStep}",
                    "CACHE_DIR=${currentCiGroup}"
                ]) {
                    image.inside("-v \'${env.WORKSPACE}/.optimize/${currentCiGroup}:${env.WORKSPACE}/optimize\'") {
                        sh "node scripts/functional_tests.js --config test/functional/config.js --include ${currentCiGroup}"
                    }
                }
                sh "rm -rf ${env.WORKSPACE}/.optimize/${currentCiGroup}"
            }
        }
```
Jenkins job test was run to confirm stable results. The past 24 test jobs passes: https://jenkins.bfs.sichend.people.aws.dev/job/Kibana/job/bfs7.7.1/
